pinot-minion-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/CustomerBasedRetentionTaskGenerator.java

```diff
@@ -0,0 +1,230 @@
+import org.apache.pinot.spi.annotations.minion.TaskGenerator;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableTaskConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.utils.TimeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@TaskGenerator
+public class CustomerBasedRetentionTaskGenerator implements PinotTaskGenerator{
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CustomerBasedRetentionTaskGenerator.class);
+  public static final String COLUMNS_TO_CONVERT_KEY = "columnsToConvert";
+  public static final int MAX_SEGMENTS_PER_TASK = 32;
+
+  private ClusterInfoAccessor _clusterInfoAccessor;
+  private HelixPropertyStore<ZNRecord> propertyStore;
+
+  @Override
+  public void init(ClusterInfoAccessor clusterInfoAccessor) {
+    _clusterInfoAccessor = clusterInfoAccessor;
+  }
+
+  @Override
+  public String getTaskType() {
+    return TASK_TYPE;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+
+    for (TableConfig tableConfig : tableConfigs) {
+      String offlineTableName = tableConfig.getTableName();
+
+      LOGGER.info("Start generating task configs for table: {} for task: {}", offlineTableName, TASK_TYPE);
+
+      // Only generate tasks for OFFLINE tables
+      if (tableConfig.getTableType() != TableType.OFFLINE) {
+        LOGGER.warn("Skip generating task: {} for non-OFFLINE table: {}", TASK_TYPE, offlineTableName);
+        continue;
+      }
+
+      // Only schedule 1 job of this type (this optimisation may be moved downstream later on)
+      Map<String, TaskState> incompleteTasks =
+          TaskGeneratorUtils.getIncompleteTasks(TASK_TYPE, offlineTableName, _clusterInfoAccessor);
+      if (!incompleteTasks.isEmpty()) {
+        LOGGER
+            .warn("Found incomplete tasks: {} for same table: {}. Skipping task generation.", incompleteTasks.keySet(),
+                offlineTableName);
+        continue;
+      }
+
+      TableTaskConfig tableTaskConfig = tableConfig.getTaskConfig();
+      Preconditions.checkNotNull(tableTaskConfig);
+      Map<String, String> taskConfigs = tableTaskConfig.getConfigsForTaskType(TASK_TYPE);
+      Preconditions.checkNotNull(taskConfigs, "Task config shouldn't be null for Table: {}", offlineTableName);
+
+      // Get customer retention config
+      Map<String ,String> customerRetentionConfigMap = getCustomerRetentionConfig();
+      String customerRetentionConfigMapString = customerRetentionConfigMap.keySet().stream()
+          .map(key -> key + "=" + customerRetentionConfigMap.get(key))
+          .collect(Collectors.joining(", ", "{", "}"));
+      SortedSet<String> sortedDistinctRetentionPeriods = getSortedDistinctRetentionPeriods(customerRetentionConfigMap);
+
+      // Get customer ID
+      String customerId = getCustomerId();
```

Comment:
@laxmanchekka Can we move this out of the loop ? These will be same for all tables right ?